estransport/connection.go

```diff
@@ -0,0 +1,280 @@
+		cp.metrics.Unlock()
+	}
+
+	return &cp
+}
+
+// Next returns the connection from pool.
+//
+func (cp *singleConnectionPool) Next() (*Connection, error) {
+	return cp.connection, nil
+}
+
+// Remove is a no-op for single connection pool.
+//
+func (cp *singleConnectionPool) Remove(c *Connection) error {
+	return nil
+}
+
+// Next returns a connection from pool, or an error.
+//
+func (cp *roundRobinConnectionPool) Next() (*Connection, error) {
+	var c *Connection
+
+	cp.Lock()
+	defer cp.Unlock()
+
+	// Return next live connection
+	if len(cp.live) > 0 {
+		cp.curr = (cp.curr + 1) % len(cp.live)
+		return cp.live[cp.curr], nil
+
+	} else if len(cp.dead) > 0 {
+		// Try to resurrect a dead connection if no live connections are available
+		if cp.debugLogger != nil {
+			cp.debugLogger.Log("Resurrecting connection ")
+		}
+		c, cp.dead = cp.dead[len(cp.dead)-1], cp.dead[:len(cp.dead)-1] // Pop item
+		c.Lock()
+		if cp.debugLogger != nil {
+			cp.debugLogger.Log(c.URL.String() + "\n")
+		}
+		c.markAsLive()
+		cp.live = append(cp.live, c)
+		c.Unlock()
+
+		return c, nil
+	}
+
+	return nil, errors.New("no connection available")
+}
+
+// Remove removes a connection from the pool.
+//
+func (cp *roundRobinConnectionPool) Remove(c *Connection) error {
+	cp.Lock()
+	defer cp.Unlock()
+
+	c.Lock()
+
+	if c.Dead {
+		if cp.debugLogger != nil {
+			cp.debugLogger.Logf("Already removed %s\n", c.URL)
+		}
+		c.Unlock()
+		return nil
+	}
+
+	if cp.debugLogger != nil {
+		cp.debugLogger.Logf("Removing %s...\n", c.URL)
+	}
+	c.markAsDead()
+	cp.scheduleResurrect(c)
+	c.Unlock()
+
+	// Push item to dead list and sort slice by number of failures
+	cp.dead = append(cp.dead, c)
+	sort.Slice(cp.dead, func(i, j int) bool {
+		c1 := cp.dead[i]
+		c2 := cp.dead[j]
+		c1.Lock()
+		c2.Lock()
+		defer c1.Unlock()
+		defer c2.Unlock()
+
+		res := c1.Failures > c2.Failures
+		return res
+	})
+
+	// Check if connection exists in the list. Return nil if it doesn't,
+	// because another goroutine might have already removed it.
+	index := -1
+	for i, conn := range cp.live {
+		if conn == c {
+			index = i
+		}
+	}
+	if index < 0 {
+		return nil
+	}
+
+	// Remove item; https://github.com/golang/go/wiki/SliceTricks
+	copy(cp.live[index:], cp.live[index+1:])
+	cp.live = cp.live[:len(cp.live)-1]
+
+	if cp.enableMetrics {
+		cp.metrics.Lock()
+		cp.metrics.dead = cp.dead
+		cp.metrics.live = cp.live
+		cp.metrics.Unlock()
+	}
+
+	return nil
+}
+
+// Resurrect removes the connection from the dead list and adds it to the pool.
+//
+// TODO(karmi): Add a pluggable strategy as argument, eg. "optimistic", "ping".
+//
+func (cp *roundRobinConnectionPool) Resurrect(c *Connection) error {
+	cp.Lock()
+	defer cp.Unlock()
+
+	c.Lock()
+	defer c.Unlock()
+
+	if !c.Dead {
+		if cp.debugLogger != nil {
+			cp.debugLogger.Logf("Already resurrected %s\n", c.URL)
+		}
+		return nil
+	}
+
+	if cp.debugLogger != nil {
+		cp.debugLogger.Logf("Resurrecting %s\n", c.URL)
+	}
+
+	c.markAsLive()
```

Comment:
I think it is worth to optimize for the good path. Is it feasible to do a healthcheck here on the resurrected connection, and if successful call `markAsHealthy`? This way you could get rid of any locking in the _good_ case of using a connection https://github.com/elastic/go-elasticsearch/pull/95/files#diff-8fce2da81ff9e3289d4a5515a242448bR263-R269.

Reply:
With a "healthcheck" you mean something like calling `HEAD /` on the node, ie. a "ping"? That's definitely something I was thinking about, but that would need a pluggable "strategy" for the resurrection, as outlined in the TODO comment, because I don't want to hardcode the "ping" into the implementation.

Reply:
That makes sense. Totally fair to do this at a later point if necessary.